### DISCUSSION TITLE
custom admin UI dashboard

### DISCUSSION THREAD

**Discussion by MustafaM257** [medusajs/medusa#6324] [Q&A]
Hi everyone! I'm currently working on a project using Medusa, and I've hit a bit of a roadblock. The admin UI is located within the node_modules, and unfortunately, I don't have direct access to it for customization purposes.

I'm looking to customize the admin dashboard but want to ensure I'm following the best practices recommended by the Medusa library. My main questions are:

What is the recommended approach for customizing the Medusa admin dashboard when direct access to the node_modules directory is not feasible?

Mainly I wanna do this in order to edit the forms, products fields change doesn't reflect on the UI

**Comment by ZeeshanMehdiDev** [comment]
Admin Widgets are custom React components that developers create to be injected into predetermined injection zones across the Medusa Admin dashboard.

Widgets allow you to customize the admin dashboard by providing merchants with new features. For example, you can add a widget on the order details page that shows payment details retrieved from Stripe.

This guide explains the available injection zones and how to create an admin widget.

[https://docs.medusajs.com/admin/widgets](url)

here you will find how to customize medusajs admin dashboard by creating widgets:
![image](https://github.com/medusajs/medusa/assets/119069286/2d4f3dc7-4e2a-4f7e-bfc2-9e6baec90ca8)

**Reply by NicolasGorga** [reply]
@Zeeshan-Mehdi110 Although one can add widgets to existing components or create new ones all together, one cannot currently edit the existing dashboard, to for example, show / hide left hand side menu items based on user roles or navigate to a custom page after login, like '/pending-verification' when a Vendor first signs up in a multi-vendor marketplace environment

**Reply by devtony01** [reply]
@Zeeshan-Mehdi110 Please answer. It's very important

**Reply by liamcsdev97** [reply]
@Zeeshan-Mehdi110 Is there any realistic way to replace or update an existing component without completely copying the admin ui src code and manually changing it? The injection zones are great for plugin developers and allows them to ship premade ui components with their libraries, but when it comes to actual in project source code, they don't provide much utility and are fairly limiting. I personally want to replace the variants section component with my own which has extended utilities. But the only way I can think of doing that is by cloning the admin ui source code into a standalone project and completely ditching the admin ui in the backend. This of course would mean that ui components shipped by plugins will no longer be accessible and I will be forced to copy them as well, or recreate my own from scratch. Or, the only other alternative I can think of is replacing the entire route, cloning the relevant components and recreating the route from scratch. I think it's very important to allow developers to replace existing components in the admin ui, or at the very least allow them  to replace organisms (which seem to be the top level component, and therefore won't affect other components in the source). Maybe just replicate the admin ui directory structure in the backend and merge it in with the base, overwritting existing files accordingly. So if I want to replace the options modal in /components/organisms/product-variants-section/options-modal.tsx, I can create the following file in my backend to replace that components: /admin/components/organisms/product-variants-section/options-modal.tsx. This functionality of overwritting components should be prohibited in plugins, as that would clearly break things. But the source project should have the ability to replace these components at their own risk. Otherwise the admin-ui is a nice default, but is a ticking time bomb. You will simply be forced to move away from it the second you need to replace something outside an injection zone.

**Reply by nozhaa** [reply]
any response ?

**Comment by ZeeshanMehdiDev** [comment]
@nozhaa To customize the Medusa admin dashboard, you can create Admin Widgets that allow you to inject custom React components into specific zones. Unfortunately, as my experience, modifying existing components directly isn't supported, so using widgets is the best approach.

**Reply by nozhaa** [reply]
@Zeeshan-Mehdi110 Using widgets is limited to actions like inserting content before or after certain elements. However, I need the ability to add or hide buttons on an existing page, which is not supported by widgets.

**Reply by NicolasGorga** [reply]
Hey @nozhaa I'll share the steps i followed, based on an article by Adil, an active member of the MedusaJS community, regarding fully customizing the Admin panel for V1. Basically the key is to fork the medusajs/admin (plugin package) and the medusajs/admin-ui (react package with actual UI) and to point you Medusa project, to your foroked version of the medusajs/admin plugin, which inturn, references your forked medusajs/admin-ui project.

- Clone @medusajs/admin and @medusajs/admin-ui
- Install dependencies in the @medusajs/admin-ui only for now
- Make your changes to @medusajs/admin-in the ui folder
- Build and publish to npm with your custom name
- Replace all the occurences inside the @medusajs/admin about @medusajs/admin-ui with your @my-custom/admin-ui npm published package
- Build it and publish your new @my-custom/admin npm package
- Install the previously publisehd package @my-custom/admin that represents the custom admin plugin that will points out to your custom admin-ui
- Update your medusa-config.js file to match your new package name instead of the Medusa.js one
- Install patch-package and go to node_modules/@medusajs/medusa/dist/commands/utils/resolve-admin-cli.js
- Replace the hardcoded value @medusajs/admin to @my-custom/admin plugin name
- Run yarn patch-package @medusajs/medusa, this will creates a patch that will be used instead of the one in the node_modules
- Add postinstall script that will run patch-package whenever someone run yarn install, it will patch even after installs
- Run yarn dev

Worthy to mention that while developing, its preferable to avoid publishing the packages to npm, as everytime you change something you would have to re publish. The key here is to use symlinks, between medusajs project > forked admin plugin > forked admin ui.

Once you are happy and ready to publish, you replace the symlinks references in the projects with your actual published npm package for both admin projects, in your main Medusa project.

I have been using this successfully, both locally and now published to npm, since i am testing deployments on various cloud providers. Beware that the medusajs/admin-ui project is HUGE and can be really overwhelming. 

The Medusa team told me V2 approach to customizing Admin is less cumbersome than with V1, but for all of us using V1 this i believe is the right way to go (and what has worked for me so far). 

Hope this helps you and others trying to customize Admin far beyong what widgets and pages allow!

**Reply by JovanJevtic** [reply]
if i'm correct this works for versions below medusa v2?

**Reply by NicolasGorga** [reply]
My reply is for v1, i haven't tried v2 yet, but there are a LOT of changes, so probably a different approach would have to be taken. A member of the team told me that one of their intentions with v2 is to allow for customizations like this to be less cumbersome, but haven't had the time to see if this is the case yet. Maybe someone has tried it and can share their experience?

**Reply by SalahAdDin** [reply]
No documentation about it yet?

**Reply by JovanJevtic** [reply]
I'm still trying. I have to trasnlate the admin dashboard and remove english language in my case. For now I achieved it by cloning the whole medusajs/medusa repo and than going to packages/admin/dashboard/i18n/... and added my language translations and configured other neccessery i18n stuff. Then I succeed to run it localy. Then I created a npm package (dashboard is root directory) and inside my main medusa backend project in package.json added `"@medusajs/dashboard": "npm:medusa-dashboard@^2.0.322",` (medusa-dashboard is the name of my npm package), but that did not change anything when running the project. The next step I'm trying is adding this inside the medusa-config.ts file `... modules: {
    "@medusajs/dashboard": {
      resolve: "medusa-dashboard",
    },
  },`  but when I try to run it I run into following issue: error:   Could not resolve module: @medusajs/dashboard. Error: Cannot find module 'virtual:medusa/links'
Require stack:
- C:\Users\ovojo\code\next-tests\longi\longi\node_modules\medusa-dashboard\dist\app.js
- C:\Users\ovojo\code\next-tests\longi\longi\node_modules\@medusajs\utils\dist\common\dynamic-import.js. Any thoughts. I'm aware that I may be doing nonsense things...

**Reply by JovanJevtic** [reply]
Update: I succeeded replacing @medusajs/dashboard by my own npm package inside medusa backend (v2.0.1) project. When I previously said that "but that did not change anything when running the project" it was because @medusajs/admin-bundler still used original package. Just for now I went straight into node_modules/@medusajs/admin-bundler/package.json and also did this`"@medusajs/dashboard": "npm:medusa-dashboard@^2.0.322"`, then I ran build but I ran into module not found errors and had to install @medusajs/ui-preset. Then I ran build again and it built successfully. After that i ran my medusa project and saw my admin dashboard with my translations. So I'm wondering if forking @medusajs/admin-bundler and replacing it in root package.json is the only solution or is anyone having a better idea? FORGOT TO MENTION: I deleted the stuff inside the medusa-config.ts that I mentioned in previous comment

**Reply by SalahAdDin** [reply]
@JovanJevtic Is it a bug? it would be nice you share the whole procedure related to it.

**Reply by tirocoder** [reply]
> Update: I succeeded replacing @medusajs/dashboard by my own npm package inside medusa backend (v2.0.1) project. When I previously said that "but that did not change anything when running the project" it was because @medusajs/admin-bundler still used original package. Just for now I went straight into node_modules/@medusajs/admin-bundler/package.json and also did this`"@medusajs/dashboard": "npm:medusa-dashboard@^2.0.322"`, then I ran build but I ran into module not found errors and had to install @medusajs/ui-preset. Then I ran build again and it built successfully. After that i ran my medusa project and saw my admin dashboard with my translations. So I'm wondering if forking @medusajs/admin-bundler and replacing it in root package.json is the only solution or is anyone having a better idea? FORGOT TO MENTION: I deleted the stuff inside the medusa-config.ts that I mentioned in previous comment

Is this sustainable? How can we manage the new updates that will come to @medusajs/dashboard with this method? https://github.com/medusajs/medusa/discussions/6324#discussioncomment-10104352 as mentioned here, isn't this method an unsustainable method for general admin UI development? I would be happy if friends who have experienced this issue before write their opinions.

**Reply by SalahAdDin** [reply]
Isn't admin a module? we could create modules to custom the admin.

**Reply by OmkarHalpatrao** [reply]
Does any one know the injection zones to add something in sidebar.

**Comment by Op-Pautu** [comment]
[this was addressed here](https://docs.perseides.org/guides/v2/customize-admin-ui/standalone)

**Reply by pavelbaltov** [reply]
following this extraction of dashboard to a standalone app, I am wondering if there is an effective way to benefit from coming dashboard related changes to the medusa repo. does anyone have a wise approach to this problem?

**Reply by Op-Pautu** [reply]
well I don't know what exactly you want to do, but your medusa backend is simply for the logic, workflows, api routes etc. while admin dashboard is UI related. you can call api endpoints from dashboard to backend

**Reply by pavelbaltov** [reply]
maybe my question could have been a bit confusing and misleading. i was wondering how to get the new changes from the contributers related to the dashboard effectivley, now that the dashboard is separated from the main repo.

**Reply by Op-Pautu** [reply]
might be a bit tedious but you could manually extract the dashboard folder from the latest release I think.

**Reply by Alishan786a** [reply]
> might be a bit tedious but you could manually extract the dashboard folder from the latest release I think.

Hello @Op-Pautu when i run it in dev mode it works fine when i create its build for production it show me empty page . Have any idea why this happening?

**Reply by SalahAdDin** [reply]
> [this was addressed here](https://docs.perseides.org/guides/v2/customize-admin-ui/standalone)

Not available anymore.

**Comment by Alishan786a** [comment]
admin panel should be separate like store-front so that developers can easily customize it. As now i want to change navbar but there is not any option .It is requested to contributers to make if separat.

**Reply by Op-Pautu** [reply]
read my comment above

**Reply by Alishan786a** [reply]
i clone the project and copy the dashboard folder to another place and i create a file with name ".yarnrc.yml"  with content "nodeLinker: node-modules" as mention in document i when i run this command "yarn install" it give me this error

Usage Error: The nearest package directory (/home/alishan/work/medusa/dashboard) doesn't seem to be part of the project declared in /home/alishan.

- If /home/alishan isn't intended to be a project, remove any yarn.lock and/or package.json file there.
- If /home/alishan is intended to be a project, it might be that you forgot to list work/medusa/dashboard in its workspace configuration.
- Finally, if /home/alishan is fine and you intend work/medusa/dashboard to be treated as a completely separate project (not even a workspace), create an empty yarn.lock file in it.

$ yarn install [--json] [--immutable] [--immutable-cache] [--check-cache] [--inline-builds] [--mode #0]

**Reply by Alishan786a** [reply]
thanks @Op-Pautu i just create yarn.lock empty file this  fixed the issue.
can you please help me with plugins how to add plugins . i add a plugin but its now working and as of now i separate the admin panel now where i install plugin in this separat admin panel or in server? once again thanks for previous support.

**Reply by Op-Pautu** [reply]
well, you just customize it like any react project. go to the component where u wanna add "plugins". there's no automatic way to add plugins or extensions you just do it manually

**Reply by Alishan786a** [reply]
Okay thanks. @Op-Pautu

**Reply by Alishan786a** [reply]
I have another question  in store frontend there is a way to register via email can I can it to phone. Is it possible?

**Reply by Op-Pautu** [reply]
u mean like otp auth? it is possible yeah but u need to setup some stuff

**Reply by Alishan786a** [reply]
@Op-Pautu hi once again,

I have another question that currently images are stored on file system on the server and is it good for production or not. If not then tell me can I use cloudinary for it.

**Reply by Op-Pautu** [reply]
you can try if your app is relatively small. but it's quite limited if you want to scale in the future. you can use their recommended s3 module or if u have to use cloudinary you can follow the guide:
https://docs.medusajs.com/resources/references/file-provider-module

**Reply by Alishan786a** [reply]
Okay thank you so much

**Reply by SalahAdDin** [reply]
> you can try if your app is relatively small. but it's quite limited if you want to scale in the future. you can use their recommended s3 module or if u have to use cloudinary you can follow the guide: https://docs.medusajs.com/resources/references/file-provider-module

do we have one for working with Cloudinary?

**Reply by Alishan786a** [reply]
i did try to add cloudinary and its begin to work

**Reply by Alishan786a** [reply]
hello @Op-Pautu!
 now i stuck at this issue i don't know why this happend. It's a frontend of medusa store.

![image](https://github.com/user-attachments/assets/ead1f569-fdff-413e-a090-396d1a0b84b8)
i have a dev server i create a build on my dev server and start it  its working on 
http://localhost:8001
and i have binded a domain with it https://babyfitter.ammag.tech
every thing is working but on https://babyfitter.ammag.tech/pk/account page is keep crashing initially it load and then crash but is working on http://localhost:8001/pk/account .
And its works well  http://localhost:8001 and  https://babyfitter.ammag.tech when i run it in dev mode. issue happend on production.

**Reply by Op-Pautu** [reply]
I don't have enough context to give you an answer. maybe remove .next folder, clear browser cache, or both and see

**Reply by Alishan786a** [reply]
@Op-Pautu sorry for disturbance i am facing issue with exporting products in medusa when export it i got the transaction it when i open notification bar i got below message. please help me to sort out the issue.
![image](https://github.com/user-attachments/assets/88b1f00d-2c7d-418f-80f3-a51f7c12c642)

**Reply by Op-Pautu** [reply]
there's an endpoint being called when you export products. u can see network tab when you click the button and debug

**Reply by Alishan786a** [reply]
> there's an endpoint being called when you export products. u can see network tab when you click the button and debug

thanks, i fixed the issue i was using custome file system that was the issue.

**Reply by Alishan786a** [reply]
@Op-Pautu hello there, i just upgrade 2.6.1 to 2.7.1 version of medusa i just changer it version package.json file and re install with npm i i am getting some issue with database i have mush data in database what i need more step which i am missing

**Reply by Op-Pautu** [reply]
did u update everything in devDependencies too?

**Reply by Alishan786a** [reply]
> did u update everything in devDependencies too?

yes i did only medusa dependencies and devDependencies.
here is error which i face i think it about db?

![image](https://github.com/user-attachments/assets/011cbc3c-75af-49cb-aab5-a6ad62947579)

**Reply by Op-Pautu** [reply]
run migration? not sure other than that

**Reply by Alishan786a** [reply]
> run migration? not sure other than that

is this command
npx medusa db:migrate

**Reply by Op-Pautu** [reply]
yeah

**Reply by Alishan786a** [reply]
@Op-Pautu i want to integrate NMI payment in medusa , can i do this? And how. Thank for your support

**Reply by Op-Pautu** [reply]
@Alishan786a there's a guide on the docs for setting up custom payment provider

**Reply by Alishan786a** [reply]
> @Alishan786a there's a guide on the docs for setting up custom payment provider

Okay thank you very much

**Comment by pepijn-vanvlaanderen** [comment]
We resolved customizing the admin dashboard and keep it maintainable in a turborepo, you can use a setup like this:
1. Download latest medusa version from source repo, extract `packages/admin/dashboard`
2. Create your own mirror repo with the dashboard contents
3. Fork the mirror for each of your projects
4. Add the fork into the main project monorepo: `git submodule add <admin-repo-url> packages/medusa-admin-dashboard`
3. Link the local dashboard package to your `apps/medusa` through the package.json: `"@medusajs/dashboard": "workspace:^"`
4. Make sure to update the medusa config so it does not load admin in development mode:
```typescript
module.exports = defineConfig({
  admin: {
    disable: process.env.NODE_ENV === 'development',
  },
  ...
})
```
5. Make sure your turbo.json contains the following so it will ensure building the admin dashboard on turbo builds:
```json
"build#medusa": {
  "dependsOn": ["^build"],
  "inputs": ["$TURBO_DEFAULT$", ".env*"],
  "outputs": [".medusa/**"]
}
```

### Committing changes
1. Navigate to the submodule folder: `cd packages/medusa-admin-dashboard`
2. Commit your changes
3. Go to root project and commit changes for the monorepo

### CI build
Note: you need to clone your project with submodules, i.e. git clone --depth=1 --recurse-submodules <main-repo-url> .

### Sync upstream
1. When Medusa admin dashboard has a new release you download latest medusa version from source repo, extract `packages/admin/dashboard` and update the mirror repo.
2. Now in you forked repos you can merge upstream changes

**Reply by pepijn-vanvlaanderen** [reply]
To make it complete for local development when you also want to build custom widgets or routes from inside your Medusa project (recommended to not alter the core dashboard to much). You then also need to add this env, otherwise they will not show:
```
VITE_MEDUSA_PROJECT=/<absolute-path-to-project-root>/apps/medusa/src/admin
```

**Reply by nishit-g** [reply]
I am unable to see the plugins in custom dashboard, currently building one custom plugin locally (following the docs).

**Comment by radoslaw-sz** [comment]
For everyone who is interested - my company is working on pluggable version of Medusa Admin - it will be open-source.

https://github.com/user-attachments/assets/1345ae44-f90a-4391-9171-714e40750173

**Reply by Sheikh-Faraz** [reply]
Hi this seems interesting where can I get to know more about it

**Reply by radoslaw-sz** [reply]
Still not yet released - we will update.
But work is ongoing with new features as in showcase :)


https://github.com/user-attachments/assets/5d1f190d-2200-4514-848f-dd18a33a2d92

**Reply by Sheikh-Faraz** [reply]
It looks but I have some questions, is it going to be free, is there any limitation in customization or no like we can customize anything, how long can we expect to come

**Reply by radoslaw-sz** [reply]
it will be open-source. There will be much more places and the whole framework which will let you inject your components or pages etc. wherever you want.
I hope in 2 weeks the very first version will be available.

**Reply by SalahAdDin** [reply]
Not release yet

**Comment by radoslaw-sz** [comment]
For everyone who is interested in high customization of Medusa - we launched new project called Nocto which tries to fixes it - https://www.noctojs.com/

**Reply by Sheikh-Faraz** [reply]
Hey bro the link is not working

**Reply by SalahAdDin** [reply]
> For everyone who is interested in high customization of Medusa - we launched new project called Nocto which tries to fixes it - https://www.noctojs.com/

To fix what? If it is a bug and not an intended behavior, it should be better to open a PR on the project.

**Reply by Sheikh-Faraz** [reply]
The link is working but I had to use Vpn for it. Is this because of a region thing ?

**Reply by radoslaw-sz** [reply]
hi @Sheikh-Faraz  - sorry, there was a DNS problem. It shall work now.

**Reply by jackhuynh95** [reply]
hi @radoslaw-sz , that is very helpful to kick-off and customize all owned places in the codebase continuously. 
In my use case, have been reaching out it with `turbo` mono-repos

**Reply by SalahAdDin** [reply]
> hi @radoslaw-sz , that is very helpful to kick-off and customize all owned places in the codebase continuously. In my use case, have been reaching out it with `turbo` mono-repos

Tutorial?

**Reply by SalahAdDin** [reply]
it was archived

**Comment by radoslaw-sz** [comment]
hi @SalahAdDin  - maybe wrong word, just improve. We decided to go with the fork due to significant changes.
hi @Sheikh-Faraz  - that's strange, could you please tell me your region?

**Reply by SalahAdDin** [reply]
> hi @SalahAdDin  - maybe wrong word, just improve. We decided to go with the fork due to significant changes.
> hi @Sheikh-Faraz  - that's strange, could you please tell me your region?

You mean you are forking Medusa?

**Comment by fatidian1** [comment]
No wonder why devs are starting their own forks of the project, because owners of medusa are ignoring 50% of issues (they are auto closed after 30 days of inactivity) and you can't modify even the easiest things in the medusa admin panel. You want to change max file upload size? Sorry, it's hardcoded in react component 🤣 Good job guys! The easiest way to modify anything would be to write your own admin panel that uses medusa.js api.

**Comment by Op-Pautu** [comment]
Guys, I linked it before but for visibility here's the [guide ](https://docs.perseides.org/guides/v2/customize-admin-ui/standalone)again.

**Reply by SalahAdDin** [reply]
> Guys, I linked it before but for visibility here's the [guide ](https://docs.perseides.org/guides/v2/customize-admin-ui/standalone)again.

It is not available anymore.
